src/GitHub.Api/UI/TreeBase.cs

```diff
@@ -0,0 +1,302 @@
+                                }
+                            }
+                        }
+                        else
+                        {
+                            isActive = treeData.IsActive;
+                            treeNodeTreeData = treeData;
+                        }
+
+                        isSelected = selectedNodePath != null && nodePath == selectedNodePath;
+                        AddNode(nodePath, label, i + displayRootLevel, isFolder, isActive, nodeIsHidden,
+                            nodeIsCollapsed, isSelected, treeNodeTreeData);
+                    }
+                }
+            }
+        }
+
+        protected abstract IEnumerable<string> GetCollapsedFolders();
+
+        protected void AddNode(string path, string label, int level, bool isFolder, bool isActive, bool isHidden,
+            bool isCollapsed, bool isSelected, TData? treeData)
+        {
+            var node = CreateTreeNode(path, label, level, isFolder, isActive, isHidden, isCollapsed, treeData);
+
+            SetNodeIcon(node);
+            Nodes.Add(node);
+
+            if (isSelected)
+            {
+                SelectedNode = node;
+            }
+        }
+
+        protected void Clear()
+        {
+            OnClear();
+            Nodes.Clear();
+            SelectedNode = null;
+        }
+
+        protected abstract void RemoveCheckedNode(TNode node);
+
+        protected abstract void AddCheckedNode(TNode node);
+
+        protected abstract TNode CreateTreeNode(string path, string label, int level, bool isFolder, bool isActive,
+            bool isHidden, bool isCollapsed, TData? treeData);
+
+        protected abstract void OnClear();
+
+        protected abstract void SetNodeIcon(TNode node);
+
+        protected void ToggleNodeVisibility(int idx, TNode node)
+        {
+            var nodeLevel = node.Level;
+            node.IsCollapsed = !node.IsCollapsed;
+            idx++;
+            for (; idx < Nodes.Count && Nodes[idx].Level > nodeLevel; idx++)
+            {
+                Nodes[idx].IsHidden = node.IsCollapsed;
+                if (Nodes[idx].IsFolder && !node.IsCollapsed && Nodes[idx].IsCollapsed)
+                {
+                    var level = Nodes[idx].Level;
+                    for (idx++; idx < Nodes.Count && Nodes[idx].Level > level; idx++)
+                    { }
+
+                    idx--;
+                }
+            }
+
+            if (SelectedNode != null && SelectedNode.IsHidden)
+            {
+                SelectedNode = node;
+            }
+        }
+
+        protected void ToggleNodeChecked(int idx, TNode node)
+        {
+            var isChecked = false;
+
+            switch (node.CheckState)
+            {
+                case CheckState.Mixed:
+                case CheckState.Empty:
+                    node.CheckState = CheckState.Checked;
+                    isChecked = true;
+                    break;
+
+                case CheckState.Checked:
```

Comment:
You don't need to `isChecked = false;` here?

Reply:
It is set to false by default a few lines up.

Reply:
Duh..